#VRML_SIM R2023a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/appearances/protos/TireRubber.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/appearances/protos/OsbWood.proto"

WorldInfo {
  basicTimeStep 16
}
Viewpoint {
  orientation -0.319336334094465 -0.17636603896360475 0.9310850262072721 4.0915321976617385
  position 2.487055950657228 0.10797958649459476 1.0653685934065968
  follow "robot_clone"
  followSmoothness 0.01
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
  contactMaterial "floor"
  floorSize 7 4.55
  wallThickness 0.1
}
Robot {
  translation 1.42 -0.5 0.0389797
  rotation 0 1 0 0
  children [
    Camera {
      translation -0.09 0 0
      rotation 0 0 1 3.14159
      width 300
      height 150
    }
    Group {
      children [
        Transform {
          translation 0.05 0 0
          children [
            DEF pi Shape {
              appearance PBRAppearance {
                baseColor 0 0 0
              }
              geometry Box {
                size 0.09 0.051 0.03
              }
            }
          ]
        }
        Transform {
          translation 0 0 -0.02
          children [
            DEF base Shape {
              appearance PBRAppearance {
                baseColor 0 0 0
              }
              geometry Box {
                size 0.2 0.06 0.02
              }
            }
          ]
        }
      ]
    }
    DEF front_right_hinge Hinge2Joint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor -0.0645468 0.04 -0.0267332
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      device [
        DEF FRServo RotationalMotor {
          name "front_right_servo"
        }
      ]
      endPoint Solid {
        translation -0.0645468 0.04 -0.0267332
        rotation 1 0 0 1.5708
        children [
          DEF front_left_wheel_shape Shape {
            appearance TireRubber {
            }
            geometry Cylinder {
              height 0.01
              radius 0.0125
            }
          }
        ]
        name "solid(2)"
        contactMaterial "tyre"
        boundingObject USE front_left_wheel_shape
        physics Physics {
        }
      }
    }
    DEF front_left_hinge Hinge2Joint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor -0.0645456 -0.04 -0.0261826
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      device [
        DEF FLServo RotationalMotor {
          name "front_left_servo"
        }
      ]
      endPoint Solid {
        translation -0.0645456 -0.04 -0.0261826
        rotation 0.9999961310822082 -0.001967200136269051 -0.001966708986824902 4.712383331178661
        children [
          DEF front_right_wheel_shape Shape {
            appearance TireRubber {
            }
            geometry Cylinder {
              height 0.01
              radius 0.0125
            }
          }
        ]
        name "solid(3)"
        contactMaterial "tyre"
        boundingObject USE front_right_wheel_shape
        physics Physics {
        }
      }
    }
    DEF rear_right_hinge HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 -1 0
        anchor 0.0754532 0.04 -0.0267332
      }
      device [
        RotationalMotor {
          name "right_wheel_motor"
          acceleration 5
          maxVelocity 100
        }
      ]
      endPoint Solid {
        translation 0.0754532 0.04 -0.0267332
        rotation 1 0 0 1.5708
        children [
          DEF rear_right_wheel_shape Shape {
            appearance TireRubber {
            }
            geometry Cylinder {
              height 0.01
              radius 0.0125
            }
          }
        ]
        contactMaterial "tyre"
        boundingObject USE rear_right_wheel_shape
        physics Physics {
        }
      }
    }
    DEF rear_left_hinge HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 -1 0
        anchor 0.0754531317196739 -0.04000000620125117 -0.026733221149208525
      }
      device [
        DEF left_wheel_motor RotationalMotor {
          name "left_wheel_motor"
          acceleration 5
          maxVelocity 100
        }
      ]
      endPoint Solid {
        translation 0.0754531317196739 -0.04000000620125117 -0.026733221149208525
        rotation 0.9999961310822082 -0.001967200136269051 -0.001966708986824902 4.712383331178661
        children [
          DEF rear_left_wheel_shape Shape {
            appearance TireRubber {
            }
            geometry Cylinder {
              height 0.01
              radius 0.0125
            }
          }
        ]
        name "solid(1)"
        contactMaterial "tyre"
        boundingObject USE rear_left_wheel_shape
        physics Physics {
        }
      }
    }
  ]
  boundingObject Transform {
    translation 0 0 -0.005
    children [
      Box {
        size 0.2 0.06 0.05
      }
    ]
  }
  physics Physics {
  }
}
Robot {
  translation 1.77 -1.05 0.0389797
  children [
    Camera {
      translation -0.09 0 0
      rotation 0 0 1 3.14159
      width 300
      height 150
    }
    Group {
      children [
        Transform {
          translation 0.05 0 0
          children [
            Shape {
              appearance PBRAppearance {
                baseColor 0.333333 0.333333 1
                metalness 0.8
              }
              geometry Box {
                size 0.09 0.051 0.03
              }
            }
          ]
        }
        Transform {
          translation 0 0 -0.02
          children [
            DEF base Shape {
              appearance PBRAppearance {
                baseColor 0.333333 0.333333 1
                metalness 0.8
              }
              geometry Box {
                size 0.2 0.06 0.02
              }
            }
          ]
        }
      ]
    }
    DEF front_right_hinge HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor -0.0645468 0.04 -0.0267332
      }
      device [
        DEF FRServo RotationalMotor {
          name "front_right_servo"
        }
      ]
      endPoint Solid {
        translation -0.0645468 0.04 -0.0267332
        children [
          HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 1 0
            }
            endPoint Solid {
              rotation 0 1 0 0
              children [
                DEF wheel_shape Transform {
                  rotation 1 0 0 -1.5707953071795862
                  children [
                    Shape {
                      appearance TireRubber {
                      }
                      geometry Cylinder {
                        height 0.01
                        radius 0.0125
                      }
                    }
                  ]
                }
              ]
              boundingObject USE wheel_shape
              physics Physics {
              }
            }
          }
        ]
        boundingObject Box {
          size 0.01 0.01 0.01
        }
        physics Physics {
        }
      }
    }
    DEF front_left_hinge HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor -0.0645468 -0.04 -0.0267332
      }
      device [
        DEF FLServo RotationalMotor {
          name "front_left_servo"
        }
      ]
      endPoint Solid {
        translation -0.0645468 -0.04 -0.0267332
        children [
          HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 1 0
            }
            endPoint Solid {
              rotation 0 1 0 0
              children [
                DEF wheel_shape Transform {
                  rotation 1 0 0 -1.5707953071795862
                  children [
                    Shape {
                      appearance TireRubber {
                      }
                      geometry Cylinder {
                        height 0.01
                        radius 0.0125
                      }
                    }
                  ]
                }
              ]
              boundingObject USE wheel_shape
              physics Physics {
              }
            }
          }
        ]
        name "solid(1)"
        boundingObject Box {
          size 0.01 0.01 0.01
        }
        physics Physics {
        }
      }
    }
    DEF rear_right_hinge HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 -1 0
        anchor 0.0754532 0.04 -0.0267332
      }
      device [
        RotationalMotor {
          name "right_wheel_motor"
          acceleration 5
          maxVelocity 100
        }
      ]
      endPoint Solid {
        translation 0.0754532 0.04 -0.0267332
        rotation 0 -1 0 0
        children [
          DEF wheel_shape Transform {
            rotation 1 0 0 -1.5707953071795862
            children [
              Shape {
                appearance TireRubber {
                }
                geometry Cylinder {
                  height 0.01
                  radius 0.0125
                }
              }
            ]
          }
        ]
        name "solid(2)"
        contactMaterial "tyre"
        boundingObject USE wheel_shape
        physics Physics {
        }
      }
    }
    DEF rear_left_hinge HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 -1 0
        anchor 0.0754532 -0.04 -0.0267332
      }
      device [
        RotationalMotor {
          name "left_wheel_motor"
          acceleration 5
          maxVelocity 100
        }
      ]
      endPoint Solid {
        translation 0.0754532 -0.04 -0.0267332
        rotation 0 -1 0 0
        children [
          DEF wheel_shape Transform {
            rotation 1 0 0 -1.5707953071795862
            children [
              Shape {
                appearance TireRubber {
                }
                geometry Cylinder {
                  height 0.01
                  radius 0.0125
                }
              }
            ]
          }
        ]
        name "solid(3)"
        contactMaterial "tyre"
        boundingObject USE wheel_shape
        physics Physics {
        }
      }
    }
  ]
  name "robot_clone"
  boundingObject Transform {
    translation 0 0 -0.005
    children [
      Box {
        size 0.2 0.06 0.05
      }
    ]
  }
  physics Physics {
  }
  controller "car_controller"
}
DEF Track Transform {
  children [
    Solid {
      translation 1.19 -2.05 0.02
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(2)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.01 -2.05 0.02
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(1)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -1.21 -2.05 0.02
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(3)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -2.23826 -1.61746 0.02
      rotation 0 0 1 -0.785395307179586
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(4)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -2.96198 -0.6752 0.02
      rotation 0 0 1 -1.047195307179586
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(5)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -3.26698 0.441746 0.02
      rotation 0 0 1 1.5708
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(6)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -2.84278 1.45277 0.02
      rotation 0 0 1 -2.3561953071795863
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(7)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -1.82317 1.87238 0.02
      rotation 0 0 -1 8.23753e-07
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(8)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.441295 1.51579 0.02
      rotation 0 0 1 -0.785395307179586
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.6 0.006 0.05
          }
        }
      ]
      name "solid(9)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.0170328 -0.0963247 0.02
      rotation 0 0 1 -0.785395307179586
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.6 0.006 0.05
          }
        }
      ]
      name "solid(11)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.958482 -1.03777 0.02
      rotation 0 0 1 1.5708
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.61 0.006 0.05
          }
        }
      ]
      name "solid(24)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation 0.395365 -0.453634 0.02
      rotation 0 0 -1 0.64
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.5 0.006 0.05
          }
        }
      ]
      name "solid(16)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation 1.99704 -1.83922 0.02
      rotation 0 0 1 0.785401
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.6 0.006 0.05
          }
        }
      ]
      name "solid(12)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation 1.99704 -0.821114 0.02
      rotation 0 0 1 2.3562
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.6 0.006 0.05
          }
        }
      ]
      name "solid(13)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation 2.20704 -1.32969 0.02
      rotation 0 0 1 -1.5707953071795862
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.6 0.006 0.05
          }
        }
      ]
      name "solid(14)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.93594 1.80006 0.02
      rotation 0 0 -1 0.261797
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 0.6 0.006 0.05
          }
        }
      ]
      name "solid(10)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.23 0.71 0.02
      rotation 0 0 1 1.5708
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      boundingObject USE track1-bounding
    }
    Solid {
      translation 1.19 -0.6 0.02
      rotation 0 0 -1 -5.307179586466759e-06
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(15)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation 0.7 -1.34 0.02
      rotation 0 0 -1 -5.307179586466759e-06
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(17)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.5 -1.34 0.02
      rotation 0 0 -1 -5.307179586466759e-06
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(19)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -1.61052 -1.03948 0.02
      rotation 0 0 1 -0.523595307179586
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(20)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -2.14982 -0.13448 0.02
      rotation 0 0 1 1.5708
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(21)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -1.54982 0.46552 0.02
      rotation 0 0 1 -3.1415853071795863
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(22)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation -0.95982 -0.13448 0.02
      rotation 0 0 -1 1.57079
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(23)"
      boundingObject USE track1-bounding
    }
    Solid {
      translation 1.19 -0.6 0.02
      rotation 0 0 -1 -5.307179586466759e-06
      children [
        DEF track1-bounding Shape {
          appearance OsbWood {
          }
          geometry Box {
            size 1.2 0.006 0.05
          }
        }
      ]
      name "solid(18)"
      boundingObject USE track1-bounding
    }
  ]
}
